src/database.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the DatabaseObserver class to handle database relation and state."""
+
+import typing
+
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+from ops.charm import CharmBase
+from ops.framework import Object
+
+from constants import DATABASE_NAME
+
+class DatabaseService(Object):
+    """The Database relation observer."""
+
+    def __init__(self, charm: CharmBase, relation_name: str):
+        """Initialize the oserver and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        self.relation_name = relation_name
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self.relation_name,
+            database_name=DATABASE_NAME,
+        )
+
+    def get_relation_data(self) -> typing.Dict[str, str]:
+        """Get database data from relation.
+
+        Returns:
+            Dict: Information needed for setting environment variables.
+            Returns default if the relation data is not correctly initialized.
+        """
+        default = {
+            "POSTGRES_USER": "",
+            "POSTGRES_PASSWORD": "",
+            "POSTGRES_HOST": "",
+            "POSTGRES_PORT": "",
+            "POSTGRES_DB": "",
+        }
```

Comment:
Shouldn't this be a pydantic model, this is leaking the environment into the database service which is not a good practice

Reply:
Yes, I took it from discourse-k8s, though I will refactor it to a pydantic model, way better.